### ISSUE TITLE
Debian packaging

### ISSUE THREAD

**Issue by cinemast** [nordlicht/nordlicht#38]
Hi!

I saw your project at 32c3 and liked it very much. Would you be interested in having your package in Debian (and therefore also in official Ubuntu repositories)?

I haven't looked at the source code / licenses yet, so if you are ok with that I will review your code maybe suggest some chnages and package it for Debian.

Greetings
Peter

**Comment by blinry** [comment]
Hey Peter, that would be pretty awesome! There is a preliminary version of a Debian package created by @rohieb at the [debian](/nordlicht/nordlicht/tree/debian) branch, you might want to look at that.

You're also very welcome to review the code and suggest improvements! Especially regarding the C API, I have never done something like this.

For such an "official" release I would also like to work towards a stable "1.0.0" version. Most importantly, the *spectrogram* style does not work wich every audio codec yet, see #26.

Thanks!

**Comment by cinemast** [comment]
Hey!

The files at the debian branch look mostly fine and only require some minor tweaks. For me the most important thing is to keep the ABI and API as stable as possible. Do you know how to manage this?

The API looks pretty straightforward and simple, I like that. You might also like to take a look at the following article: http://lucumr.pocoo.org/2013/8/18/beautiful-native-libraries/

So should I wait for the next big 1.0 release?

**Comment by cinemast** [comment]
There might come more comments with suggestions :)

For now, could you please consider signing your releases with GPG?

**Comment by rohieb** [comment]
> The files at the debian branch look mostly fine and only require some minor
> tweaks.

Oh, I'm interested what I missed in my packaging work :)

I also considered packaging the mpv integration, but when I packaged it, mpv was
very fresh in Debian (I think only version 0.2?) and didn't have the new Lua API
yet. That has changed, and I guess the mpv integration can now also be included.

 - Roland

**Comment by cinemast** [comment]
Could you give some details about the mpv integration? What exactly is that? 

I'll drop the final changes to the package, as soon as I am done. Its basically just cosmetics, and some changes that happened inside the Debian project (e.g. automatic debug packages).

**Comment by cinemast** [comment]
The following files might be nice having in the project:

- LICENSE - basically what README already contains
- CHANGELOG - a list of changes that occured in each version.

@blinry: Could you please provide this in the next release?

**Comment by cinemast** [comment]
Could you please make downloading the test file optional (e.g. by having some variable that can be passed via cmake call)?

**Comment by blinry** [comment]
LICENSE/CHANGELOG: Sure! :-) Will come in a few days!

Signed release: Like this? https://github.com/nordlicht/nordlicht/releases/tag/v0.4.2

Optional test file: Yeah, sounds useful. Maybe even to make it the default not to download the file.

As for the mpv integration, it's basically a Lua script which generates a nordlicht of each file opened in the video player mpv, plus a bit of user interaction. See [the integrations page](https://nordlicht.github.io/integrations/#mpv) for details, or directly read [the script](https://github.com/nordlicht/nordlicht/blob/master/utils/mpv-nordlicht.lua). I honestly think it's not really ready for packaging yet, as it's quite hacky at the moment :P It's rather a proof-of-concept, which I personally use.

**Comment by rohieb** [comment]
On 04.01.2016 23:15, Peter Spiess-Knafl wrote:
> Could you give some details about the mpv integration? What exactly is that?

It's a Lua script (utils/mpv-nordlicht.lua) that can be used as a plugin for mpv
(with the --script= option). It generates a nordlicht when a video file is
loaded and shows it when the file is being played. I just tested it with mpv
0.14 from unstable, it works fine :)

> I'll drop the final changes to the package, as soon as I am done. Its basically
> just cosmetics, and some changes that happened inside the Debian project (e.g.
> automatic debug packages).

Ah cool, nice to hear that automatic debug packages are now a thing :)

 - Roland

**Comment by cinemast** [comment]
@blinry : 
- The signature looks good, thanks! It would be greate if you could also sign all future releases.
- There are missing license files for cheat.h and cmake/FindPopt and cmake/FindFFmpeg.
- There are lots of warnings from flawfinder (forget about the warnings for cheat.h). Could you take a look at that and maybe fix them?
http://paste.debian.net/360319/

Thanks again for your collaboration on all that.

**Comment by blinry** [comment]
@cinemast:

Some clarification regarding the downloading of the test file: I checked this, and on my setup, that only happens when executing `make check`. For the test suite, having some kind of video file is essential, as you cannot usefully test anything without an input file.

Does the file get downloaded for you when issuing a normal `make`? Or do you think this is a problem even when wanting to run the test suite? In the latter case, maybe we could include a tiny video file with the project, or even autogenerate one using FFmpeg? Would that help?

In the latest master, I've included a `CHANGELOG.md` and a `LICENSE.md`, have a look. I'll also look at the flawfinder warnings today.

Thanks for all your suggestions, I'm always happy to learn! :-)

**Comment by blinry** [comment]
I checked all of flawfinder's warnings. I replaced an `sprintf` call with `snprintf` (but the buffer size was already calculated correctly before), but found none of the other warnings to be relevant. I think it's okay to do a `system` call in the testsuite and check whether some files exist there. An "attacker" could merely make some tests fail.

Also, there are warnings about `open` calls (about symlinks, redirection) – but the user has already full control over the file path to be opened and can do with it what they want anyway. There are warnings about old incorrect implementations of `snprintf`, but I think it's the standard just to use it.

Please object if you see any risks.

**Comment by cinemast** [comment]
Hi!

At a quick overlook, your changes look very fine. The license stuff is really very important. Thanks for fixing that.

I checked about the file download, its OK if it doesn't get downloaded automatically. I will just disable the test suite during the build.

Also good job on the changelog. It makes packaging so much easier in the future. I am not an security expert myself and therefore must take a deeper look into this. For now its fine.

So it would be great if you could create a new release with a new signature, to have the latest version packaged.

Thanks for your collaboration.

Greetings.

**Comment by blinry** [comment]
Peter, I created a new signed release for you – v0.4.3! Let us know how it goes!

**Comment by blinry** [comment]
In cc86cd362844700fced80ebcf443bb8f8a82c961, I also fixed the symbol visibility [mentioned by Sebastian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=809923#24).

**Comment by cinemast** [comment]
Hi!

I provided another pullrequest (#46), fixing the Multiarch issue. I think its worth having upstream. Also the library versioning is quite important. 

After that it would be great having another upstream release. Which I can than again package for Debian. I think we are very close now.

Thank you and also @rohieb  for your support in this.

**Comment by blinry** [comment]
@cinemast There you go, v0.4.4 :-) Thanks for all your suggestions and all your work on this!

**Comment by cinemast** [comment]
Thank you! I already propsed a new version to Debian-Multimedia. So probably It will also get team-maintained there, which is a good thing.

http://anonscm.debian.org/cgit/collab-maint/nordlicht.git/log/

**Comment by rohieb** [comment]
Yay :-)

```
$ apt-cache policy libnordlicht0
libnordlicht0:
  Installed: 0.3.5+git20151005.bb21ed2-1
  Candidate: 0.4.4-1
  Version table:
     0.4.4-1 170
        170 http://ftp.de.debian.org/debian sid/main amd64 Packages
 *** 0.3.5+git20151005.bb21ed2-1 100
        100 /var/lib/dpkg/status
```

**Comment by cinemast** [comment]
It currently fails on arm architectures. There is a proposed fix: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=813101

Could you take a look at that?

**Comment by blinry** [comment]
Yay, we're in unstable :) Good work!

I addressed the arm problem in 2e33a66. I'm also noticed that the arm build fails in [this log](https://buildd.debian.org/status/fetch.php?pkg=nordlicht&arch=arm64&ver=0.4.4-1&stamp=1454082292), but I'm not quite sure why - is it because of the warning about ftruncate? I also fixed that (in 3f77aff).

I will take a one or two days to round off some corners, then make a new release.
